go/vt/vttablet/tabletserver/vstreamer/snapshot_conn.go

```diff
@@ -217,3 +217,43 @@ func GetBinlogRotationThreshold() int64 {
 func SetBinlogRotationThreshold(threshold int64) {
 	atomic.StoreInt64(&binlogRotationThreshold, threshold)
 }
+
+func (conn *snapshotConn) startSnapshotAllTables(ctx context.Context) (gtid string, err error) {
+	lockConn, err := mysqlConnect(ctx, conn.cp)
+	if err != nil {
+		return "", err
+	}
+	// To be safe, always unlock tables, even if lock tables might fail.
+	defer func() {
+		_, err := lockConn.ExecuteFetch("unlock tables", 0, false)
+		if err != nil {
+			log.Warning("Unlock tables failed: %v", err)
+		} else {
+			log.Infof("Tables unlocked")
+		}
+		lockConn.Close()
+	}()
+
+	log.Infof("Locking all tables")
+	if _, err := lockConn.ExecuteFetch("FLUSH TABLES WITH READ LOCK", 1, false); err != nil {
+		log.Infof("Error locking all tables")
+		return "", err
+	}
+	mpos, err := lockConn.PrimaryPosition()
+	if err != nil {
+		return "", err
+	}
+
+	// Starting a transaction now will allow us to start the read later,
+	// which will happen after we release the lock on the table.
+	if _, err := conn.ExecuteFetch("set transaction isolation level repeatable read", 1, false); err != nil {
+		return "", err
+	}
```

Comment:
This is not needed when using `START TRANSACTION WITH CONSISTENT SNAPSHOT` but it also doesn't hurt.